src/UI/Implementation/Component/Button/Toggle.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+/* Copyright (c) 2018 Thomas Famula <famula@leifos.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Implementation\Component\Button;
+
+use ILIAS\UI\Component as C;
+use ILIAS\UI\Component\Signal;
+use ILIAS\UI\Implementation\Component\ComponentHelper;
+use ILIAS\UI\Implementation\Component\JavaScriptBindable;
+use ILIAS\UI\Implementation\Component\Triggerer;
+
+class Toggle extends Button implements C\Button\Toggle {
+	use ComponentHelper;
+	use JavaScriptBindable;
+	use Triggerer;
+
+	/**
+	 * @var string|null
+	 */
+	protected $action_deactivated;
+
+	public function __construct($label, $action, $action_deactivated)
+	{
+		parent::__construct($label, $action);
+
+		$this->checkArg(
+			"action_deactivated",
+			is_string($action_deactivated) || $action_deactivated instanceof Signal,
+			$this->wrongTypeMessage("string or Signal", gettype($action_deactivated))
+		);
+		if (is_string($action_deactivated)) {
+			$this->action_deactivated = $action_deactivated;
+		}
+		else {
+			$this->action_deactivated = null;
+			$this->setTriggeredSignal($action_deactivated, "click");
+		}
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function getActionDeactivated()
+	{
+		if ($this->action_deactivated !== null) {
+			return $this->action_deactivated;
+		}
+		$triggered_click_signals = $this->triggered_signals["click"];
+		if ($triggered_click_signals === null) {
+			return [];
+		}
+		return array_map(
+			function($ts) { return $ts->getSignal(); },
+			$triggered_click_signals
+		);
```

Comment:
Could you explain what this here is supposed to accomplish exactly. Looks cryptic to me, but this might also be a problem in the Signals or my general understanding of these.

Reply:
This is simply copied from the `getAction()` method from the `Button.php` class. My idea was that the action for the deactivated state, including the Signal, will be handled like the default action (in this case the action for the activated state).

Reply:
Ok, this actually my code (in Button.php) =)
This

```
if ($this->action_deactivated !== null) {
	return $this->action_deactivated;
}
```
accounts for the fact, that a button could either use a string for the action or use a signal. The interface assures, that either the one or the other is used. The storage for the string is provided by the button itself while the signals are handled by the trait `Triggerer`.

```
$triggered_click_signals = $this->triggered_signals["click"];
if ($triggered_click_signals === null) {
	return [];
}
```
would actually be better placed in `Triggerer` (maybe as `getTriggeredSignalsFor`) since it accounts for an internal storage implementation of the trait.

```
return array_map(
	function($ts) { return $ts->getSignal(); },
	$triggered_click_signals
);
```
The triggered signals array actually contains a tuple (signal,event) (named `TriggeredSignal`) not the signals themselves. This steps grabs the signal from that tuple. This imho would be better located in `Triggerer`  as well.

Reply:
Performed the refactoring, see [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1109).